weavetest/decorators.go

```diff
@@ -0,0 +1,74 @@
+package weavetest
+
+import "github.com/iov-one/weave"
+
+// Decorator is a mock implementation of the weave.Decorator interface.
+//
+// Set CheckErr or DeliverErr to force error response for corresponding method.
+// If error attributes are not set then wrapped handler method is called and
+// its result returned.
+// Each method call is counted. Regardless of the method call result the
+// counter is incremented.
+type Decorator struct {
+	checkCall int
+	// CheckErr if set is returned by the Check method before calling
+	// the wrapped handler.
+	CheckErr error
+
+	deliverCall int
+	// DeliverErr if set is returned by the Deliver method before calling
+	// the wrapped handler.
+	DeliverErr error
+}
+
+var _ weave.Decorator = (*Decorator)(nil)
+
+func (d *Decorator) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx, next weave.Checker) (weave.CheckResult, error) {
+	d.checkCall++
+
+	if d.CheckErr != nil {
+		return weave.CheckResult{}, d.CheckErr
+	}
+	return next.Check(ctx, db, tx)
+}
+
+func (d *Decorator) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx, next weave.Deliverer) (weave.DeliverResult, error) {
+	d.deliverCall++
+
+	if d.DeliverErr != nil {
+		return weave.DeliverResult{}, d.DeliverErr
+	}
+	return next.Deliver(ctx, db, tx)
+}
+
+func (d *Decorator) CheckCallCount() int {
+	return d.checkCall
+}
+
+func (d *Decorator) DeliverCallCount() int {
+	return d.deliverCall
+}
+
+func (d *Decorator) CallCount() int {
+	return d.checkCall + d.deliverCall
+}
+
+func Decorate(h weave.Handler, d weave.Decorator) weave.Handler {
```

Comment:
Looks a lot like app.ChainDecorators.

Not sure if this duplication is good or bad

Reply:
Good catch! I did not notice that.

It was called [`Wrap`](https://github.com/iov-one/weave/blob/master/x/helpers.go#L83) before so I have ported it. Should I replace it? I can see both pros and cons.

Reply:
I guess it is fine.
I don't want dependencies between weavetest and app also
Maybe leave a comment about the other version for clarity